include/swift/AST/AutoDiff.h

```diff
@@ -231,6 +232,18 @@ void getSubsetParameterTypes(IndexSubset *indices, AnyFunctionType *type,
                              SmallVectorImpl<Type> &results,
                              bool reverseCurryLevels = false);
 
+/// "Constrained" derivative generic signatures require all differentiability
+/// parameters to conform to the `Differentiable` protocol.
+///
+/// Returns the "constrained" derivative generic signature given:
+/// - An original SIL function type.
+/// - Differentiability parameter indices.
+/// - A possibly "unconstrained" derivative generic signature.
+GenericSignature
+getConstrainedDerivativeGenericSignature(SILFunctionType *originalFnTy,
```

Comment:
Why isn't there a version for `ASTFunctionType`?

Reply:
Good question - I guess an `AnyFunctionType` version hasn't been necessary for correctness.

I'm pretty sure the `SILFunctionType` version is necessary in the [differentiation transform](https://github.com/apple/swift/pull/29621/files#diff-4ede198777b59748c377cdd17c444e83R567). So I don't believe an `AnyFunctionType` version can replace the `SILFunctionType` version.